src/nginx/module.cc

```diff
@@ -69,6 +69,21 @@ ngx_esp_request_ctx_s::ngx_esp_request_ctx_s(ngx_http_request_t *r,
       backend_time(-1) {
   ngx_memzero(&wakeup_event, sizeof(wakeup_event));
   if (lc && lc->esp) {
+    std::string destination_url;
+    if (lc->esp->ReWriteURL(ngx_str_to_std(r->unparsed_uri),
+                            &destination_url)) {
+      ngx_str_copy_from_std(r->pool, destination_url, &r->unparsed_uri);
+
+      ngx_str_copy_from_std(r->pool, destination_url.substr(
+                                         0, destination_url.find_first_of('?')),
+                            &r->uri);
```

Comment:
My impression is: r.uri is unescaped.   But I am not sure.  Could you double check it?

Reply:
That is correct. r->uri is unescaped. 

r->uri=/api/shelves/2+ -$/
r->unparsed_uri=/api/shelves/2+%20-$///?key=AIzaSyCfvOENA9MbRupfKQau2X_l8NGMVWF_byI
r->request_line=GET /api/shelves/2+%20-$///?key=AIzaSyCfvOENA9MbRupfKQau2X_l8NGMVWF_byI HTTP/1.1

Called ngx_esp_rewrite_url using r->url, instead of implementing unescape.